File: shared/src/main/java/org/dwallach/calwatch/EventWrapper.java
/*
 * CalWatch
 * Copyright (C) 2014 by Dan Wallach
 * Home page: http://www.cs.rice.edu/~dwallach/calwatch/
 * Licensing: http://www.cs.rice.edu/~dwallach/calwatch/licensing.html
 */
package org.dwallach.calwatch;

import android.graphics.Paint;

/**
 * Created by dwallach on 8/25/14.
 */
public class EventWrapper {
    private final static String TAG = "EventWrapper";
    private WireEvent wireEvent;
    private PathCache pathCache;
    private Paint paint, greyPaint, lowBitPaint;
    private int minLevel, maxLevel;


    public EventWrapper(WireEvent wireEvent) {
        this.wireEvent = wireEvent;
        this.pathCache = new PathCache();
        this.paint = PaintCan.getCalendarPaint(wireEvent.displayColor);
        this.greyPaint = PaintCan.getCalendarGreyPaint(wireEvent.displayColor);
        this.lowBitPaint = PaintCan.get(true, true, PaintCan.colorBlackFill);
        this.minLevel = this.maxLevel = 0;  // fill this in later on...
    }

    public WireEvent getWireEvent() {
        return wireEvent;
    }

    public PathCache getPathCache() {
        return pathCache;
    }

    public Paint getPaint(boolean ambientLowBit, boolean ambientMode) {
        if(ambientMode)
            if(ambientLowBit)
                return lowBitPaint;
            else
                return greyPaint;
        else
            return paint;
    }

    public int getMinLevel() { return minLevel; }

    public void setMinLevel(int minLevel) { this.minLevel = minLevel; }

    public int getMaxLevel() { return maxLevel; }

    public void setMaxLevel(int maxLevel) { this.maxLevel = maxLevel; }

    public boolean overlaps(EventWrapper e) {
        return this.wireEvent.startTime < e.wireEvent.endTime && e.wireEvent.startTime < this.wireEvent.endTime;
    }

    public String toString() {
        return wireEvent.startTime + "->" + wireEvent.endTime + " color(" + Integer.toHexString(wireEvent.displayColor) + "), levels(" + minLevel + "," + maxLevel + ")";
    }
}

File: shared/src/main/kotlin/org/dwallach/calwatch/EventWrapper.kt
/*
 * CalWatch
 * Copyright (C) 2014 by Dan Wallach
 * Home page: http://www.cs.rice.edu/~dwallach/calwatch/
 * Licensing: http://www.cs.rice.edu/~dwallach/calwatch/licensing.html
 */
package org.dwallach.calwatch

import android.graphics.Paint
import android.graphics.Path

class EventWrapper(val wireEvent: WireEvent) {
    /**
     * The first time this event is rendered, it will be rendered to a Path, so subsequent calls to
     * render it will go much faster.
     */
    var path: Path? = null
    private val paint = PaintCan.getCalendarPaint(wireEvent.displayColor)
    private val greyPaint = PaintCan.getCalendarGreyPaint(wireEvent.displayColor)
    private val lowBitPaint = PaintCan[true, true, PaintCan.colorLowBitCalendarFill]
    var minLevel: Int = 0
    var maxLevel: Int = 0

    fun getPaint(ambientLowBit: Boolean, ambientMode: Boolean) = when {
        ambientMode && ambientLowBit -> lowBitPaint
        ambientMode -> greyPaint
        else -> paint
    }

    fun overlaps(e: EventWrapper) =
            this.wireEvent.startTime < e.wireEvent.endTime && e.wireEvent.startTime < this.wireEvent.endTime

    override fun toString() =
            "%d -> %d, color(%08x), levels(%d,%d)"
                    .format(wireEvent.startTime, wireEvent.endTime, wireEvent.displayColor, minLevel, maxLevel)
}
